src/codegen/compile.ml

```diff
@@ -4226,35 +3817,57 @@ module Blob = struct
       BigNum.from_word32 env
     )
 
-  let vanilla_lit env s =
-    Tagged.shared_static_obj env Tagged.Blob StaticBytes.[
-      I32 (Int32.of_int (String.length s));
-      Bytes s;
-    ]
-
-  let lit env s = compile_unboxed_const (vanilla_lit env s)
-
-  let lit_ptr_len env s =
-    compile_unboxed_const (Int32.add ptr_unskew (E.add_static env StaticBytes.[Bytes s])) ^^
-    compile_unboxed_const (Int32.of_int (String.length s))
+  (* unskewed target address, data offset, and length on stack *)
+  let load_static_data env s =
+    let segment_index = E.add_static env StaticBytes.[Bytes s] in
+    G.i (MemoryInit (nr segment_index))
 
   let alloc env =
     E.call_import env "rts" "alloc_blob" ^^
     (* uninitialized blob payload is allowed by the barrier *)
     Tagged.allocation_barrier env
 
-  let unskewed_payload_offset env = Int32.(add ptr_unskew (mul Heap.word_size (header_size env)))
-
+  let unskewed_payload_offset env = Int32.(add ptr_unskew (mul Heap.word_size header_size))
+  
   let payload_ptr_unskewed env =
     Tagged.load_forwarding_pointer env ^^
     compile_add_const (unskewed_payload_offset env)
 
+  let constant env payload =
+    E.SharedObject (fun env -> 
+      let blob_length = Int32.of_int (String.length payload) in
+      let (set_new_blob, get_new_blob) = new_local env "new_blob" in
+      compile_unboxed_const blob_length ^^ alloc env ^^ set_new_blob ^^
+      get_new_blob ^^ payload_ptr_unskewed env ^^ (* target address *)
+      compile_unboxed_const 0l ^^ (* data offset *)
+      compile_unboxed_const blob_length ^^ (* data length *)
+      load_static_data env payload ^^
+      get_new_blob
+    )
+
+  let lit env payload =
+    Tagged.share_constant env (constant env payload)
+
   let as_ptr_len env = Func.share_code1 Func.Never env "as_ptr_size" ("x", I32Type) [I32Type; I32Type] (
     fun env get_x ->
       get_x ^^ payload_ptr_unskewed env ^^
       get_x ^^ len env
     )
 
+  let lit_ptr_len env s =
+    lit env s ^^
+    as_ptr_len env
+
+  let load_data_segment env segment_index data_length =
```

Comment:
Propably it could be refactored, e.g. by using `load_data_segment` from `constant` (the other way around). The difference here is, that `constant` directly allocates the data segment, while `load_data_segment` accesses an already defined data segment with a content defined at a later time (`create_global_type_descriptor`).